LoRa improvements

Mani-Sadhasivam opened the pull request:
1. Got rid of counter API as it is found to be not fit for upcoming LoRaWAN support.
2. Marked LoRa support as Experimental
3. Switched to new timeout API

This PR is a spinoff from #23664

Signed-off-by: Manivannan Sadhasivam <mani@kernel.org>

Mani-Sadhasivam committed:
drivers: lora: Get rid of counter support

The RTC/Counter implementation doesn't fit for the upcoming LoRaWAN
as most of the Counter drivers in Zephyr works with 1s granularity
which is not enough for LoRaWAN stack. So, k_timer calls are used in
place of Counter's alarm and k_uptime_get() APIs are used in place of
Counter's time keeping.

While at it, lets also fixup the broken alarm implementation.

Signed-off-by: Manivannan Sadhasivam <mani@kernel.org>

Mani-Sadhasivam committed:
lora: Make the LoRa support Experimental

The LoRa APIs are expected to undergo some change as the usecase get's
increased. So let's mark it as experimental until the APIs got
stabililized.

Signed-off-by: Manivannan Sadhasivam <mani@kernel.org>

andysan commented:
Thanks for pulling this out of the LoRaWAN PR, it's going the be helpful for https://github.com/zephyrproject-rtos/zephyr/pull/24616. Would it make sense to split the RTC code into a separate source file right away since you're rewriting most of it anyway? Saves me from doing it later. :)

I verified the assumptions in the new implementation against the documentation and one of the other board implementations in LoRaMAC-Node and it looks good to me.

carlescufi commented:
@andysan would you mind approving if you are happy with this?

carlescufi commented:
CC @KwonTae-young

carlescufi commented on drivers/lora/sx1276.c:
```diff
@@ -112,51 +109,61 @@ void BoardCriticalSectionEnd(uint32_t *mask)
 }
 
-void DelayMsMcu(uint32_t ms)
+void DelayMsMcu(u32_t ms)
```
`k_sleep()` will only work on threads. If you need to sleep on ISRs (hope not, because those are ms, not ns) then you need to use a `k_busy_wait()`

andysan replied:
IIRC, the ISR in the SX1276 driver uses a work queue to run the library handler in a thread, so this should be safe.

Mani-Sadhasivam replied:
@carlescufi This routine is not called by ISR, so we are safe here!

mniestroj commented on drivers/lora/sx1276.c:
```diff
@@ -112,51 +109,61 @@ void BoardCriticalSectionEnd(uint32_t *mask)
+void RtcSetAlarm(u32_t timeout)
 {
-	return 0;
+	k_timer_start(&dev_data.timer, k_ticks_to_ms_floor32(timeout),
```
New `k_timer_start()` gets `k_timeout_t` arguments instead of `s32_t`. And those are not miliseconds anymore if you disable `CONFIG_LEGACY_TIMEOUT_API` (once enough code will be converted).

Mani-Sadhasivam replied:
Will convert the whole driver to new timeout API

sslupsky replied:
Since this involves using the rtc, you may want to have a look at the discussion in #23526

Mani-Sadhasivam replied:
@sslupsky We are getting rid of RTC in LoRa. This is mainly due to the limitation of some of the RTC drivers which provide 1s granularity.

mniestroj commented on drivers/lora/sx1276.c:
```diff
@@ -112,51 +109,61 @@ void BoardCriticalSectionEnd(uint32_t *mask)
 	k_sleep(ms);
 }
 
+u32_t RtcMs2Tick(uint32_t milliseconds)
```
You are missing `RtcTick2Ms()` implementation.

Mani-Sadhasivam replied:
It was not present in initial code either so thought of adding it in LoRaWAN PR. But I can add it in this PR as well (as a separate commit though)

mniestroj reviewed (dismissed):
If I am correct then Rtc in loramac-node is used only for implementing Timer in loramac-node. So now we have:
Timer (loramac-node) <-> Rtc (loramac-node) <-> k_timer (Zephyr)

In long term we should think about implementing Timer API instead of Rtc API, which will require less bloat. This however requires changes in loramac-node repository, because it is not possible now.

andysan commented on drivers/lora/sx1276.c:
```diff
@@ -112,51 +109,61 @@ void BoardCriticalSectionEnd(uint32_t *mask)
+u32_t RtcSetTimerContext(void)
 {
-	return counter_us_to_ticks(dev_data.counter, (milliseconds / 1000));
+	saved_time = k_ms_to_ticks_ceil32(k_uptime_get_32());
```
If we want to deal with ticks anyway, why don't use the k_uptime_ticks API instead and store the saved time as an s64_t? The documentation actually states that k_uptime_get_32 uses the full precision clock, so there is normally no benefit to using the 64-bit version.

jukkar replied:
We could most probably get rid of ticks and only deal with real time. The k_timeout_t supports also us or ns timevalues.

Mani-Sadhasivam replied:
> This is really just an observation, but couldn't we just define ticks in LoRaMAC-node as being milliseconds and do away with all of the annoying time conversions? That way, tick2ms and ms2tick suddenly become unit operations. We would probably have to return k_ticks_to_ms_ceil32(1) as the minimum timeout in that case though.

The idea is not to change the LoRaMac-node repository. If we start doing this then for sure the delta between upstream and downstream will increase and I don't want to have that :) We should live with what the repository provides. If we want to change something then we _should_ do that in upstream repo and inherit here.

Mani-Sadhasivam replied:
> We could most probably get rid of ticks and only deal with real time. The k_timeout_t supports also us or ns timevalues.

Sorry no. It is like abusing the API exposed by LoRaMac-node repo. We already did that once with Rtc/timer but there we had no other choice. But for this case, since we have an API for getting the ticks, I'd like to stick to it.

sslupsky replied:
> If we want to change something then we should do that in upstream repo and inherit here.

Maybe post an issue on the Semtech repo and ask if they are receptive to integrating with zephyr?

Mani-Sadhasivam replied:
> If we want to deal with ticks anyway, why don't use the k_uptime_ticks API instead and store the saved time as an s64_t? The documentation actually states that k_uptime_get_32 uses the full precision clock, so there is normally no benefit to using the 64-bit version.

Yep, we could do that. Actually it is the right approach. Will do, thanks!

Mani-Sadhasivam replied:
> Maybe post an issue on the Semtech repo and ask if they are receptive to integrating with zephyr?

Yeah, we can do that but that is not something which will happen immediately (I mean the change there). So in the meantime, we'll stick with this.

sslupsky replied:
It will help with decisions down the road about how much to use the underlying zephyr kernel API's.

sslupsky replied:
My thought is Semtech may have an interested to make their stack available across a larger base of devices and reduce their maintenance.

Mani-Sadhasivam replied:
> My thought is Semtech may have an interested to make their stack available across a larger base of devices and reduce their maintenance.

But Zephyr is not the only user of their stack...

andysan replied:
What I meant by defining ticks as ms wouldn't require a change in the LoRaMAC code. The only change it would involve would be the way we expose ticks to the library. It's no different from a real system that just happens to define a tick as 1 ms. This would avoid the weird conversions where we go from microseconds to ticks, and then from ticks to microseconds when scheduling alarms.

If we do that, RtcMs2Tick(x) and RtcTick2Ms(x) would just return x. Calls that take a tick as an argument could just rely on a tick in LoRa being defined as 1ms, or use RtcTick2Ms to document the fact (the call would get inlined).

Mani-Sadhasivam replied:
For the record, I've switched to `k_uptime_get_32()` API

andysan reviewed (commented):
This is really just an observation, but couldn't we just define ticks in LoRaMAC-node as being milliseconds and do away with all of the annoying time conversions? That way, tick2ms and ms2tick suddenly become unit operations. We would probably have to return k_ticks_to_ms_ceil32(1) as the minimum timeout in that case though.

sslupsky commented:
Can someone comment on the minimum resources required to run this stack on zephyr?  Would the stm32l04 mcu inside the Murata lora module be able to run the lora stack on zephyr?  I am considering to build a new LoRaWAN firmware stack for the Arduino MKR WAN 1310 that runs on the Murata module.  The module currently runs a version of the Semtech AT Slave application.  However, Arduino has not updated it with the more recent 4.4.2 firmware releases from Semtech so it is terribly out of date and lacks support for time synchronization and fuota.

Mani-Sadhasivam committed:
lora: Switch to new timeout API

Get rid of legacy timeout API and move to new timeout API for LoRa.
This involves changes to API, SX1276 driver and sample application.

Signed-off-by: Manivannan Sadhasivam <mani@kernel.org>

Mani-Sadhasivam committed:
lora: sx1276: Add missing RtcTick2Ms function definition

RtcTick2Ms function is required for upcoming LoRaWAN support. Hence,
add definition for it.

Signed-off-by: Manivannan Sadhasivam <mani@kernel.org>

carlescufi commented:
@mniestroj can you please take another look?

nordic-krch commented on drivers/lora/sx1276.c:
```diff
@@ -388,8 +400,10 @@ static int sx1276_lora_recv(struct device *dev, u8_t *data, u8_t size,
-	ret = k_sem_take(&dev_data.data_sem, timeout == K_FOREVER ? K_FOREVER :
-			 timeout == K_NO_WAIT ? K_NO_WAIT : K_MSEC(timeout));
+	ret = k_sem_take(&dev_data.data_sem,
+			 K_TIMEOUT_EQ(timeout, K_FOREVER) ? K_FOREVER :
```
i don't think you need that anymore. `k_sem_take(..., timeout)` should be enough.

Mani-Sadhasivam replied:
Ack

mniestroj reviewed (approved):
LGTM
